pkg/agent/nplagent/k8s/annotations.go

```diff
@@ -0,0 +1,164 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package k8s
+
+import (
+	"context"
+	"encoding/json"
+
+	"github.com/vmware-tanzu/antrea/pkg/util/env"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/klog"
+)
+
+const NPLAnnotationStr = "npl.antrea.io"
+
+type NPLEPAnnotation struct {
+	PodPort  string `json:"Podport"`
+	NodeIP   string `json:"Nodeip"`
+	NodePort string `json:"Nodeport"`
+}
+
+func Stringify(serialize interface{}) string {
+	json_marshalled, _ := json.Marshal(serialize)
+	return string(json_marshalled)
+}
+
+func IsNodePortInAnnotation(s []NPLEPAnnotation, nodeport string) bool {
+	for _, i := range s {
+		if i.NodePort == nodeport {
+			return true
+		}
+	}
+	return false
+}
+
+func assignPodAnnotation(pod *corev1.Pod, containerPort, nodeIP, nodePort string) {
+	var err error
+	current := make(map[string]string)
+	if pod.Annotations != nil {
+		current = pod.Annotations
+	}
+
+	klog.Infof("Building annotation for pod: %s\tport: %s --> %s:%s", pod.Name, containerPort, nodeIP, nodePort)
+
+	var annotations []NPLEPAnnotation
+	// nplEP annotation exists
+	if current[NPLAnnotationStr] != "" {
+		if err = json.Unmarshal([]byte(current[NPLAnnotationStr]), &annotations); err != nil {
+			klog.Warningf("Unable to unmarshal NPLEP annotation")
+		}
+
+		if !IsNodePortInAnnotation(annotations, nodePort) {
+			annotations = append(annotations, NPLEPAnnotation{
+				PodPort:  containerPort,
+				NodeIP:   nodeIP,
+				NodePort: nodePort,
+			})
+		} else {
+			// mapping for the containerPort already exists
+			// TODO
+		}
+	} else {
+		annotations = []NPLEPAnnotation{NPLEPAnnotation{
+			PodPort:  containerPort,
+			NodeIP:   nodeIP,
+			NodePort: nodePort,
+		}}
+	}
+
+	current[NPLAnnotationStr] = Stringify(annotations)
+	pod.Annotations = current
+}
+
+func removeFromPodAnnotation(pod *corev1.Pod, containerPort string) {
+	var err error
+	current := pod.Annotations
+
+	klog.Infof("Removing annotation from pod: %s\tport: %s", pod.Name, containerPort)
+	var annotations []NPLEPAnnotation
+	if err = json.Unmarshal([]byte(current[NPLAnnotationStr]), &annotations); err != nil {
+		klog.Warningf("Unable to unmarshal NPLEP annotation")
+		return
+	}
+
+	for i, ann := range annotations {
+		if ann.PodPort == containerPort {
+			annotations = append(annotations[:i], annotations[i+1:]...)
+			break
+		}
+	}
+
+	current[NPLAnnotationStr] = Stringify(annotations)
+	pod.Annotations = current
+}
+
+// RemoveNPLAnnotationFromPods : Removes npl annotations from all pods
+func (c *Controller) RemoveNPLAnnotationFromPods() {
+	podList, err := c.KubeClient.CoreV1().Pods("").List(context.TODO(), metav1.ListOptions{})
```

Comment:
This will be expensive, in large scale, a single such call will cause considerable overhead to kube-apiserver.
Currently antrea-agent only lists Pods of this Node once on start, and it uses "resourceVersion=0" in the options to avoid reading from etcd, which caused performance issue in scale clusters. See https://github.com/vmware-tanzu/antrea/issues/1044 for details.
I think if it's one time job, it could reuse the result of https://github.com/vmware-tanzu/antrea/blob/714af4df1c826131075d4dc65d50f4c6f7049902/pkg/agent/cniserver/server.go#L606 which lists Pods of this Node only.

Reply:
Yes, this is one time operation only. I'll follow your suggestion.